api/src/Doctrine/Filter/ContentNodePeriodFilter.php

```diff
@@ -0,0 +1,82 @@
+use ApiPlatform\Core\Bridge\Doctrine\Orm\Filter\AbstractContextAwareFilter;
+use ApiPlatform\Core\Bridge\Doctrine\Orm\Util\QueryNameGeneratorInterface;
+use App\Entity\Activity;
+use App\Entity\ContentNode;
+use Doctrine\ORM\Query\Expr\Join;
+use Doctrine\ORM\QueryBuilder;
+use Doctrine\Persistence\ManagerRegistry;
+use Psr\Log\LoggerInterface;
+use Symfony\Component\HttpFoundation\RequestStack;
+use Symfony\Component\PropertyInfo\Type;
+use Symfony\Component\Serializer\NameConverter\NameConverterInterface;
+
+final class ContentNodePeriodFilter extends AbstractContextAwareFilter {
+    public const PERIOD_QUERY_NAME = 'period';
+
+    public function __construct(
+        private IriConverterInterface $iriConverter,
+        ManagerRegistry $managerRegistry,
+        ?RequestStack $requestStack = null,
+        LoggerInterface $logger = null,
+        array $properties = null,
+        NameConverterInterface $nameConverter = null
+    ) {
+        parent::__construct($managerRegistry, $requestStack, $logger, $properties, $nameConverter);
+    }
+
+    // This function is only used to hook in documentation generators (supported by Swagger and Hydra)
+    public function getDescription(string $resourceClass): array {
+        $description = [];
+        $description['period'] = [
+            'property' => self::PERIOD_QUERY_NAME,
+            'type' => Type::BUILTIN_TYPE_STRING,
+            'required' => false,
+        ];
+
+        return $description;
+    }
+
+    protected function filterProperty(
+        string $property,
+        $value,
+        QueryBuilder $queryBuilder,
+        QueryNameGeneratorInterface $queryNameGenerator,
+        string $resourceClass,
+        string $operationName = null
+    ) {
+        if (ContentNode::class !== $resourceClass) {
+            throw new \Exception("ContentNodePeriodFilter can only be applied to entities of type ContentNode (received: {$resourceClass}).");
+        }
+
+        if (self::PERIOD_QUERY_NAME !== $property) {
+            return;
+        }
+
+        // load period from query parameter value
+        $period = $this->iriConverter->getItemfromIri($value);
+
+        // generate alias to avoid interference with other filters
+        $periodParameterName = $queryNameGenerator->generateParameterName($property);
+        $rootJoinAlias = $queryNameGenerator->generateJoinAlias('root');
+        $ownerJoinAlias = $queryNameGenerator->generateJoinAlias('owner');
+        $activityJoinAlias = $queryNameGenerator->generateJoinAlias('activity');
+        $scheduleEntryJoinAlias = $queryNameGenerator->generateJoinAlias('scheduleEntry');
+
+        $rootAlias = $queryBuilder->getRootAliases()[0];
+
+        $queryBuilder
+            ->join("{$rootAlias}.root", $rootJoinAlias)
+            ->join("{$rootJoinAlias}.owner", $ownerJoinAlias)
+            ->join(Activity::class, $activityJoinAlias, Join::WITH, "{$activityJoinAlias}.id = {$ownerJoinAlias}.id")
```

Comment:
Shouldn't you technically add another constraint for the entitytype of the inheritance table? Otherwise we could have (security) bugs when a category has the same id as an activity. But okay, given roughly 900000 activities and 100000 categories (estimates based on ecamp v2) and 12-character ids using a 36 character alphabet, the chance of a collision is still only roughly 0.000002%.

Reply:
It's not possible that an activity and a category have the same id. They would both need an entry in the inheritance table (abstract_content_node_owner) with the same id, where id is the (unique) primary key of this table.